README.rst
Reading List
============

|travis| |master-coverage|  |readthedocs|

.. |travis| image:: https://travis-ci.org/mozilla-services/readinglist.svg?branch=master
    :target: https://travis-ci.org/mozilla-services/readinglist

.. |readthedocs| image:: https://readthedocs.org/projects/readinglist/badge/?version=latest
    :target: https://readinglist.readthedocs.io/en/latest/
    :alt: Documentation Status

.. |master-coverage| image::
    https://coveralls.io/repos/mozilla-services/readinglist/badge.png?branch=master
    :alt: Coverage
    :target: https://coveralls.io/r/mozilla-services/readinglist

*Reading List* is a service that aims to synchronize a list of articles URLs
between a set of devices owned by a same account.

* `Online documentation <https://readinglist.readthedocs.io/en/latest/>`_
* `Issue tracker <https://github.com/mozilla-services/readinglist/issues>`_
* `Contributing <https://readinglist.readthedocs.io/en/latest/contributing.html>`_


CHANGELOG.rst
Changelog
=========

This document describes changes between each past release.

2.1.0 (unreleased)
------------------

- Nothing changed yet.


2.0.0 (2015-07-22)
------------------

Upgraded to Cliquet 2.3+.

**Breaking changes**

- Now requires PostgreSQL 9.4+
- Endpoints now expect articles to be posted in a ``data`` attribute
- Endpoints responses now contain a ``data`` attribute
- Endpoints switched from ``If-Modified-Since`` and ``If-Unmodified-Since``
  to Etags


1.7.0 (2015-04-15)
------------------

**New feature**

- Article ``title`` now allowed to be empty or null (#250 and #253)
  (Original https://bugzilla.mozilla.org/show_bug.cgi?id=1152915)

**Bug fixes**

- Fix ``stored_on`` not being forced on creation (fixes #215)

**Internal changes**

- Enabled Coveralls with failing tests if coverage less than 100%


1.6.0 (2015-04-10)
------------------

**Deployment instructions**

Some changes were introduced in database schema. Run schema migration command
before starting the application:

::

    cliquet --ini production.ini migrate

**New features**

- Add more info in heartbeat (fixes #229)
- Clarify conflict docs (#244)
- Clarify data model docs (#247)
- Add PostgreSQL schema migration system (mozilla-services/cliquet#139)

See `every features brought by Cliquet 1.7 <https://github.com/mozilla-services/cliquet/releases/tag/1.7.0>`_

**Bug fixes**

- Fix login prompt when Basic Auth is disabled (#237)
- Fix random IndexError in load tests (#238)
- Fix smoke tests configuration reading
- Fix Heka logging format of objects (#199)
- Fix performance of record insertion using ordered index (mozilla-services/cliquet#138)
- Fix 405 errors not JSON formatted (mozilla-services/cliquet#88)

See `every bug fixes brought by Cliquet 1.7 <https://github.com/mozilla-services/cliquet/releases/tag/1.7.0>`_


1.5.0 (2015-03-30)
------------------

**New features**

- Split schema initialization from application startup, using a command-line
  tool.

::

    cliquet --ini production.ini init


**Bug fixes**

- Fix documentation about WSGI and Sentry
- Fix connection pool no being shared between cache and storage (mozilla-services/cliquet#176)
- Default connection pool size to 10 (instead of 50) (mozilla-services/cliquet#176)
- Warn if PostgreSQL session has not UTC timezone (mozilla-services/cliquet#177)

**Internal changes**

- Deprecated ``cliquet.storage_pool_maxconn`` and ``cliquet.cache_pool_maxconn``
  settings (renamed to ``cliquet.storage_pool_size`` and ``cliquet.cache_pool_size``)


1.4.0 (2015-03-27)
------------------

**New features**

- Smoke test of FxA authentication using Loads (#220)
- Mesure calls on the authentication policy (mozilla-services/cliquet#167)
- Force default pagination to 100 if not set in settings (#214)
- Add documentation about setting up Sentry loggers (#227)

**Breaking changes**

- Prefix statsd metrics with the value of `cliquet.statsd_prefix` or
  `cliquet.project_name` (mozilla-services/cliquet#162)
- `http_scheme` setting has been replaced by `cliquet.http_scheme` and
  `cliquet.http_host` was introduced (mozilla-services/cliquet#151, mozilla-services/cliquet#166)
- URL in the hello view now has version prefix (mozilla-services/cliquet#165)

**Bug fixes**

- Fix changing read position (#213)
- Fix some PostgreSQL connection bottlenecks (mozilla-services/cliquet#170)
- Pull monitoring dependencies during install (#218)

**Internal changes**

- Update of PyFxA to get it working with gevent monkey patching (mozilla-services/cliquet#168)


1.3.0 (2015-03-25)
------------------

**Deployment instructions**

Until the database schema migration system is released (mozilla-services/cliquet#139),
changes on schema have to be applied manually:

::

    ALTER FUNCTION as_epoch(TIMESTAMP) IMMUTABLE;
    CREATE INDEX idx_records_last_modified_epoch ON records(as_epoch(last_modified));
    CREATE INDEX idx_deleted_last_modified_epoch ON deleted(as_epoch(last_modified));

**New features**

- Add setting to enable to asynchronous PostgreSQL using `Psycogreen <https://pypi.python.org/pypi/psycogreen>`_.
  (*default: disabled*). See installation documentation for more details on this.
- Add ability to execute only action in loads tests using the ``LOAD_ACTION``
  environment variable. See contributing documentation for more details (#208).
- Add new load tests with several kinds of batch operations (#204)

**Bug fixes**

- Fix pagination URL in Next-page headers (fixes #210)
- Fix regression on records URL unicity when using ujson (#205)
- Fix hashing of user_id for BasicAuth (mozilla-services/cliquet#128)
- Force PostgreSQl session timezone to UTC (mozilla-services/cliquet#122)
- Make sure the `paginate_by` setting overrides the passed `limit`
  argument (mozilla-services/cliquet#129)
- Fix limit comparison under Python3 (mozilla-services/cliquet#143)
- Do not serialize using JSON if not necessary (mozilla-services/cliquet#131)
- Fix crash of classic logger with unicode (mozilla-services/cliquet#142)
- Fix crash of CloudStorage backend when remote returns 500 (mozilla-services/cliquet#142)
- Fix behaviour of CloudStorage with backslashes in querystring (mozilla-services/cliquet#142)
- Fix python3.4 segmentation fault (mozilla-services/cliquet#142)
- Add missing port in Next-Page header (mozilla-services/cliquet#147)


**Internal changes**

- Use postgres cache in loads tests (#203)
- Use ujson again, it was removed in the 1.3.2 release (#132)
- Add index for as_epoch(last_modified) (#130). Please add the following
  statements to SQL for the migration::
- Prevent fetching to many records for one user collection (#130)
- Use UPSERT for the heartbeat (#141)
- Improve tests of basic auth (#128)


1.2.0 (2015-03-20)
------------------

**New features**

- Add PostgreSQL connection pooling, with new settings
  ``cliquet.storage_pool_maxconn`` and ``cliquet.cache_pool_maxconn``
  (*Default: 50*) (mozilla-services/cliquet#112)

loadtests/loadtest/__init__.py
import json
import os
import random
import uuid

from requests.auth import HTTPBasicAuth, AuthBase
from loads.case import TestCase
from konfig import Config

ACTIONS_FREQUENCIES = [
    ('create', 20),
    ('batch_create', 50),
    ('update', 50),
    ('filter_sort', 60),
    ('read_further', 80),
    ('batch_read_further', 80),
    ('mark_as_read', 40),
    ('create_conflict', 10),
    ('update_conflict', 10),
    ('archive', 10),
    ('batch_archive', 30),
    ('delete', 10),
    ('batch_delete', 10),
    ('poll_changes', 90),
    ('list_archived', 20),
    ('list_deleted', 40),
    ('batch_count', 50),
    ('list_continuated_pagination', 80),
]


def build_article():
    suffix = uuid.uuid4().hex
    data = {
        "title": "Corp Site {0}".format(suffix),
        "url": "http://mozilla.org/{0}".format(suffix),
        "resolved_url": "http://mozilla.org/{0}".format(suffix),
        "added_by": "FxOS-{0}".format(suffix),
    }
    return data


class RawAuth(AuthBase):
    def __init__(self, authorization):
        self.authorization = authorization

    def __call__(self, r):
        r.headers['Authorization'] = self.authorization
        return r


class TestBasic(TestCase):
    def __init__(self, *args, **kwargs):
        """Initialization that happens once per user.

        :note:

            This method is called as many times as number of users.
        """
        super(TestBasic, self).__init__(*args, **kwargs)

        self.conf = self._get_configuration()

        if self.conf.get('smoke', False):
            self.random_user = "test@restmail.net"
            self.auth = RawAuth("Bearer %s" % self.conf.get('token'))
        else:
            self.random_user = uuid.uuid4().hex
            self.auth = HTTPBasicAuth(self.random_user, 'secret')

        # Create at least some records for this user
        self.nb_initial_records = random.randint(3, 100)

    def _get_configuration(self):
        # Loads is removing the extra information contained in the ini files,
        # so we need to parse it again.
        config_file = self.config['config']
        # When copying the configuration files, we lose the config/ prefix so,
        # try to read from this folder in case the file doesn't exist.
        if not os.path.isfile(config_file):
            config_file = os.path.basename(config_file)
            if not os.path.isfile(config_file):
                msg = 'Unable to locate the configuration file, aborting.'
                raise LookupError(msg)
        return Config(config_file).get_map('loads')

    def api_url(self, path):
        return "{0}/v2/{1}".format(self.server_url, path)

    def setUp(self):
        """Choose some random records in the whole collection.

        :note:

            This method is called as many times as number of hits.
        """
        nb_initial_records = self.nb_initial_records
        while nb_initial_records > 0:
            self.create()
            nb_initial_records -= 1

        resp = self.session.get(self.api_url('articles'), auth=self.auth)
        records = resp.json()['data']

        # Pick a random record
        self.random_record = random.choice(records)
        self.random_id = self.random_record['id']
        self.random_url = self.api_url('articles/%s' % self.random_id)

        # Pick another random, different
        records.remove(self.random_record)
        self.random_record_2 = random.choice(records)
        self.random_id_2 = self.random_record_2['id']
        self.random_url_2 = self.api_url('articles/%s' % self.random_id_2)

    def test_all(self):
        """Choose a random action among available, if not frequent enough,
        try again recursively.

        :note:

            This method is called as many times as number of hits.
        """
        action, percentage = random.choice(ACTIONS_FREQUENCIES)

        forced_action = os.getenv('LOAD_ACTION')
        if forced_action:
            action, percentage = forced_action, 101

        if random.randint(0, 100) < percentage:
            self.incr_counter(action)
            return getattr(self, action)()
        else:
            self.test_all()

    def _run_batch(self, body):
        resp = self.session.post(self.api_url('batch'),
                                 data=json.dumps(body),
                                 auth=self.auth,
                                 headers={'Content-Type': 'application/json'})
        self.incr_counter(resp.status_code)
        self.assertEqual(resp.status_code, 200)
        for subresponse in resp.json()['responses']:
            self.incr_counter(subresponse['status'])

    def create(self):
        data = build_article()
        resp = self.session.post(
            self.api_url('articles'),
            data=json.dumps({'data': data}),
            auth=self.auth,
            headers={'Content-Type': 'application/json'})
        self.incr_counter(resp.status_code)
        self.assertEqual(resp.status_code, 201)

    def batch_create(self):
        data = {
            "defaults": {
                "method": "POST",
                "path": "/articles"
            }
        }
        for i in range(25):
            request = {"body": {'data': build_article()}}
            data.setdefault("requests", []).append(request)

        self._run_batch(data)

    def create_conflict(self):
        data = self.random_record.copy()
        data.pop('id')
        resp = self.session.post(
            self.api_url('articles'),
            data=json.dumps({'data': data}),
            auth=self.auth,
            headers={'Content-Type': 'application/json'})
        self.incr_counter(resp.status_code)
        self.assertEqual(resp.status_code, 200)

    def filter_sort(self):
        queries = [
            [('archived', 'false')],
            [('unread', 'true'), ('archived', 'false')],
            [('_sort', '-last_modified'), ('archived', 'true')],
            [('_sort', 'title')],
            [('_sort', '-added_by,-stored_on'), ('archived', 'false')],
        ]
        queryparams = random.choice(queries)
        query_url = '&'.join(['='.join(param) for param in queryparams])
        url = self.api_url('articles?{}'.format(query_url))
        resp = self.session.get(url, auth=self.auth)
        self.incr_counter(resp.status_code)
        self.assertEqual(resp.status_code, 200)

    def _patch(self, url, data, status=200):
        body = json.dumps({'data': data})
        resp = self.session.patch(url, body, auth=self.auth)
        self.incr_counter(resp.status_code)
        self.assertEqual(resp.status_code, status)

    def update(self):
        data = {
            "title": "Some title {}".format(random.randint(0, 1)),
            "archived": bool(random.randint(0, 1)),
            "is_article": bool(random.randint(0, 1)),
            "favorite": bool(random.randint(0, 1)),
        }
        self._patch(self.random_url, data)

    def read_further(self):
        data = {
            "read_position": random.randint(0, 10000)
        }
        self._patch(self.random_url, data)

    def batch_read_further(self):
        # Get some random articles on which the batch will be applied
        url = self.api_url('articles?_limit=5&_sort=title')
        resp = self.session.get(url, auth=self.auth)
        articles = resp.json()['data']
        urls = ['/articles/{}'.format(a['id']) for a in articles]

        data = {
            "defaults": {
                "method": "PATCH",
            }
        }
        for i in range(25):
            request = {
                "path": urls[i % len(urls)],
                "body": {
                    "read_position": random.randint(0, 10000)
                }
            }
            data.setdefault("requests", []).append(request)

        self._run_batch(data)

    def mark_as_read(self):
        data = {
            "marked_read_by": "Desktop",
            "marked_read_on": 12345,
            "unread": False,
        }
        self._patch(self.random_url, data)

    def update_conflict(self):
        random_resolved_url = self.random_record_2['resolved_url']
        data = {
            "resolved_url": random_resolved_url
        }
        self._patch(self.random_url, data, status=409)

    def archive(self):
        data = {
            "archived": "true"
        }
        self._patch(self.random_url, data)

    def batch_archive(self):
        data = {
            "defaults": {
                "method": "PATCH",
                "body": {"archived": "true"}
            },
            "requests": [
                {"path": '/articles/%s' % self.random_id},
                {"path": '/articles/%s' % self.random_id_2}
            ]
        }
        self._run_batch(data)

    def delete(self):
        resp = self.session.delete(self.random_url, auth=self.auth)
        self.incr_counter(resp.status_code)
        self.assertEqual(resp.status_code, 200)

    def batch_delete(self):
        # Get some random articles on which the batch will be applied
        url = self.api_url('articles?_limit=5&_sort=title')
        resp = self.session.get(url, auth=self.auth)
        articles = resp.json()['data']
        urls = ['/articles/{}'.format(a['id']) for a in articles]

        data = {
            "defaults": {
                "method": "DELETE"
            }
        }
        for i in range(25):
            request = {"path": urls[i % len(urls)]}
            data.setdefault("requests", []).append(request)

        self._run_batch(data)

    def poll_changes(self):
        last_modified = self.random_record['last_modified']
        modified_url = self.api_url('articles?_since=%s' % last_modified)
        resp = self.session.get(modified_url, auth=self.auth)
        self.assertEqual(resp.status_code, 200)

    def list_archived(self):
        archived_url = self.api_url('articles?archived=true')
        resp = self.session.get(archived_url, auth=self.auth)
        self.assertEqual(resp.status_code, 200)

    def batch_count(self):
        data = {
            "defaults": {
                "method": "HEAD",
            },
            "requests": [
                {"path": "/articles?archived=true"},
                {"path": "/articles?is_article=true"},
                {"path": "/articles?favorite=true"},
                {"path": "/articles?unread=false"},
                {"path": "/articles?min_read_position=100"}
            ]
        }
        self._run_batch(data)

    def list_deleted(self):
        modif = self.random_record['last_modified']
        deleted_url = self.api_url('articles?_since=%s&deleted=true' % modif)
        resp = self.session.get(deleted_url, auth=self.auth)
        self.assertEqual(resp.status_code, 200)

    def list_continuated_pagination(self):
        paginated_url = self.api_url('articles?_limit=20')

        while paginated_url:
            resp = self.session.get(paginated_url, auth=self.auth)
            self.assertEqual(resp.status_code, 200)
            next_page = resp.headers.get("Next-Page")
            self.assertNotEqual(paginated_url, next_page)
            paginated_url = next_page
